[rfc] temporary custom pickle functions for maintaining BC on ConvPackedParams

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43230 [rfc] temporary custom pickle functions for maintaining BC on ConvPackedParams**
* #43086 not for land: Adding a version serialization type to ConvPackedParam

Summary:

This is an another approach to changing ConvPackedParams structure
while maintaining BC.  We add the ability to add custom pickle functions
where `__set_state__` takes a `c10::IValue` and knows how to parse it,
and define the custom pickle methods with the parsing logic on
`ConvPackedParams`.

There are various TODOs before land left in the diff, ideally we
can align if this is the approach people are OK with and implement
the TODOs after consensus.

For additional context, https://github.com/pytorch/pytorch/pull/43087
contained an alternative version of this.

Test Plan:

For now, check that saving a model with
https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be
on an old revision and loading it on this revision works.
Tests would be needed if this approach sounds good.

Reviewers:

Subscribers:

Tasks:

Tags:

vkuzo committed:
[rfc] temporary custom pickle functions for maintaining BC on ConvPackedParams

Summary:

This is an another approach to changing ConvPackedParams structure
while maintaining BC.  We add the ability to add custom pickle functions
where `__set_state__` takes a `c10::IValue` and knows how to parse it,
and define the custom pickle methods with the parsing logic on
`ConvPackedParams`.

There are various TODOs before land left in the diff, ideally we
can align if this is the approach people are OK with and implement
the TODOs after consensus.

For additional context, https://github.com/pytorch/pytorch/pull/43087
contained an alternative version of this.

Test Plan:

For now, check that saving a model with
https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be
on an old revision and loading it on this revision works.
Tests would be needed if this approach sounds good.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[rfc] temporary custom pickle functions for maintaining BC on ConvPackedParams"


Summary:

This is an another approach to changing ConvPackedParams structure
while maintaining BC.  We add the ability to add custom pickle functions
where `__set_state__` takes a `c10::IValue` and knows how to parse it,
and define the custom pickle methods with the parsing logic on
`ConvPackedParams`.

There are various TODOs before land left in the diff, ideally we
can align if this is the approach people are OK with and implement
the TODOs after consensus.

For additional context, https://github.com/pytorch/pytorch/pull/43087
contained an alternative version of this.

Test Plan:

For now, check that saving a model with
https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be
on an old revision and loading it on this revision works.
Tests would be needed if this approach sounds good.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/custom_class.h:
```diff
@@ -221,6 +221,82 @@ class class_ {
+  // version of __set_state__, which knows how to parse historical formats.
+  // TODO(before land): create an issue to delete this in a month and link here.
+  template <typename GetStateFn, typename SetStateFn>
+  class_& def_pickle_unboxed_DO_NOT_USE(
```
this should say `boxed`, my bad

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -221,6 +221,82 @@ class class_ {
+  // TODO(before land): create an issue to delete this in a month and link here.
+  template <typename GetStateFn, typename SetStateFn>
+  class_& def_pickle_unboxed_DO_NOT_USE(
+      GetStateFn&& get_state, SetStateFn&& set_state) {
```
you can simplify the code quite a bit by explicitly spelling out SetStateFn as `std::function<intrusive_ptr<CurClass>(IValue)>` and avoid templates. It's way more explicit

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -221,6 +221,82 @@ class class_ {
+                                c10::tagged_capsule<CurClass> self,
+                                SetStateArg&& arg) {
+      c10::intrusive_ptr<CurClass> classObj =
+          at::guts::invoke(set_state, std::forward<SetStateArg>(arg));
```
then you don't need invoke(), just call the std::function

dzhulgakov commented on torch/custom_class.h:
```diff
@@ -221,6 +221,82 @@ class class_ {
+      auto object = self.ivalue.toObject();
+      object->setSlot(0, c10::IValue::make_capsule(classObj));
+    };
+    defineMethod(
```
hm, how does it even work? I don't think over schema inference can handle IValue tbh. Can you print out the schema of what gets registered (i.e. `setstate_schema ` below)?

I'd expect this to be `_def_unboxed` call (actually the name is wrong - it should be `_def_boxed`). And as the schema you can construct the schema explicitly by taking the type of the argument from `getstate`.

The idea is that you want the schema for `__setsttate__` to be correct to make it invocable from Torchscript (not used right now but shall we do it ever). But the implementation can be "generic" and accept just IValue (when invoked by pickler).

Maybe I miss something and our schema actually support IValue-valued args. In which case your implementation might be ok

vkuzo committed:
Update on "[rfc] temporary custom pickle functions for maintaining BC on ConvPackedParams"


Summary:

This is an another approach to changing ConvPackedParams structure
while maintaining BC.  We add the ability to add custom pickle functions
where `__set_state__` takes a `c10::IValue` and knows how to parse it,
and define the custom pickle methods with the parsing logic on
`ConvPackedParams`.

There are various TODOs before land left in the diff, ideally we
can align if this is the approach people are OK with and implement
the TODOs after consensus.

For additional context, https://github.com/pytorch/pytorch/pull/43087
contained an alternative version of this.

Test Plan:

For now, check that saving a model with
https://gist.github.com/vkuzo/f3616c5de1b3109cb2a1f504feed69be
on an old revision and loading it on this revision works.
Tests would be needed if this approach sounds good.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov reviewed (commented):
Ah, nice, I didn't know we allowed "Any" as operator arguments :)

vkuzo commented on torch/custom_class.h:
```diff
@@ -207,16 +224,19 @@ class class_ {
-        " but expected ",
-        ser_type->repr_str());
+
+    if (validate_setstate_input_type) {
```
thanks! I squashed this PR with https://github.com/pytorch/pytorch/pull/43086 since now things are close to landable, let me move this comment there

vkuzo commented:
this was landed in https://github.com/pytorch/pytorch/pull/43086
